plugins/statuslines/lightline.nix

```
@@ -1,117 +1,206 @@
-              };
-          }
-        );
-      };
+    tabline =
+      defaultNullOpts.mkAttrsOf (with types; listOf (listOf str))
+        {
+          left = [ [ "tabs" ] ];
+          right = [ [ "close" ] ];
+        }
+        ''
+          Components placement for tabline.
+        '';
 
-      inactive = mkOption {
-        default = null;
-        description = "List of components for inactive windows.";
-        type = types.nullOr (
-          types.submodule {
-            options =
-              let
-                listType = with helpers.nixvimTypes; maybeRaw (listOf (listOf str));
-              in
-              {
-                left = helpers.mkNullOrOption listType "List of components that will show up on the left side of the bar";
+    tab = defaultNullOpts.mkAttrsOf (with types; listOf str) {
+      active = [
+        "tabnum"
+        "filename"
+        "modified"
+      ];
+      inactive = [
+        "tabnum"
+        "filename"
+        "modified"
+      ];
+    } ''A dictionary to store the tab components in each tabs.'';
 
-                right = helpers.mkNullOrOption listType "List of components that will show up on the right side of the bar";
-              };
-          }
-        );
-      };
+    mode_map =
+      defaultNullOpts.mkAttrsOf types.str
+        {
+          "n" = "NORMAL";
+          "i" = "INSERT";
+          "R" = "REPLACE";
+          "v" = "VISUAL";
+          "V" = "V-LINE";
+          "\<C-v>" = "V-BLOCK";
+          "c" = "COMMAND";
+          "s" = "SELECT";
+          "S" = "S-LINE";
+          "\<C-s>" = "S-BLOCK";
+          "t" = "TERMINAL";
+        }
+        ''
+          Mode name mappings
+        '';
+  };
 
-      modeMap = mkOption {
-        type = with types; nullOr (attrsOf str);
-        description = "Mode name mappings";
-        default = null;
-      };
+  settingsExample = {
+    colorscheme = "gruvbox";
+    component_function = {
+      gitbranch = "FugitiveHead";
+    };
+    component = {
+      charvaluehex = "0x%B";
+      lineinfo = "%3l:%-2v%<";
+    };
+    active = {
+      right = [
+        [ "lineinfo" ]
+        [ "percent" ]
+        [
+          "fileformat"
+          "fileencoding"
+          "filetype"
+          "charvaluehex"
+        ]
+      ];
+    };
+    inactive = [ ];
+    mode_map = {
+      "n" = "N";
+      "i" = "I";
+      "v" = "V";
+      "<C-v>" = "VB";
+      "<C-s>" = "SB";
     };
   };
 
-  config =
-    let
-      configAttrs = filterAttrs (_: v: v != null) {
-        inherit (cfg)
-          colorscheme
-          active
-          inactive
-          component
-          componentFunction
-          modeMap
-          ;
-      };
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-      globals.lightline = mkIf (configAttrs != { }) configAttrs;
-    };
+  callSetup = false;
+  extraConfig = cfg: { globals.lightline = cfg.settings; };
```

I think we're moving towards preferring "real" aliases for these kinda things, to preserve wrappers such as `mkDefault` if the user for whatever reason defines config in _both_ `plugins.lightline.settings` _and_ `globals.lightline`...

e.g. https://github.com/nix-community/nixvim/blob/052ee66dbb975960cc151db0b5db6a893503582a/plugins/pluginmanagers/lz-n.nix#L195

and https://github.com/nix-community/nixvim/blob/052ee66dbb975960cc151db0b5db6a893503582a/plugins/utils/firenvim.nix#L92

```suggestion
  extraConfig = cfg: {
    globals.lightline = lib.modules.mkAliasAndWrapDefsWithPriority lib.id options.plugins.lightline.settings;
  };
```

---

Thanks, I didn't realize that's what this function was. So it basically creates a 2 way binding with some sort of priority?

---

> So it basically creates a 2 way binding with some sort of priority

That's what `mkRenamedOptionModule` and `mkAliasOptionModule` do. This function only makes a one-way alias, but it does it in a way that preserves any priority wrappers used on definitions.

---

When I say "wrapped with a priority", I mean how for example `mkDefault` _wraps_ the config value as:

```nix
{
  _type = "override";
  priority = 1000;
  content = "foo";
}
```